Feature: TSCN Chunks

DasCapschen opened the pull request:
Replaces PR #236  
I did not want to rebase that... it had a lot of refactoring in it that wasn't related to TSCN chunks.  
So I rewrote it. This now also supports the new Scenario System.

THIS IS CURRENTLY UNTESTED. But all the necessary code changes should be there, apart from possibly bug fixes 😄 
Will test and fix tomorrow, hopefully 😅 

Completely replaces the `.save`, `.trackinfo` and `.scenario` files with `.tscn` files.
`jSaveModule` is no longer used anywhere.

Worlds have not yet been converted!

fixes #388 
fixes #150

HaSa1002 commented on src/project.godot:
```diff
@@ -44,6 +49,11 @@ _global_script_classes=[ {
+"base": "RailLogicSettings",
+"class": "ContactPointSettings",
+"language": "GDScript",
+"path": "res://Data/Scripts/DataClasses/contact_point_settings.gd"
```
Please rename `DataClasses` to underscore_case

DasCapschen replied:
Are you sure? All our folders are CamelCase at the moment, only files use snake_case 🤔

HaSa1002 replied:
I'd wish to change all of that but this is going to make backports very... painful

DasCapschen replied:
okay, anyways, this folder exists no more. I've moved them to Resources now, because, well, they are resources now 😄

HaSa1002 reviewed (changes requested):
A tired review. Mostly questions

DasCapschen commented on src/Data/Scripts/DataClasses/rail_logic_settings.gd:
```diff
@@ -0,0 +1,2 @@
+class_name RailLogicSettings
+extends Reference
```
Uh, I haven't done this in a while. Can Reference not be serialised? In that case, you're right, those need to be Resource and export their variables. I'll test that later today.

DasCapschen commented on src/Data/Scripts/DataClasses/route_point_despawn.gd:
```diff
@@ -0,0 +1,12 @@
+var rail_name := ""
+var distance_on_rail := 0.0
+
+func _init() -> void:
+	type = RoutePointType.DESPAWN_POINT
```
You're correct, this Type Enum is not really needed anymore, since I refactored it into classes 😅

DasCapschen commented on src/Data/Scripts/DataClasses/scenario_route.gd:
```diff
@@ -0,0 +1,214 @@
+
+func get_calculated_rail_route(world: Node) -> Array:
+	world.update_rail_connections() # why is this necessary? :(
+	var rail_route: Array = []
+
+	for i in range (size()-1):
+		# get start and end rails for calculation
+		var start_end_rails: Array = []
+		var start_direction_set: bool = false
+		var forward: bool = true
```
This is just Jeans code that I copied over :D 
I replaced the ScenarioManager and RouteManager with simply the scenario and the route.

DasCapschen commented on src/Data/Scripts/DataClasses/scenario_route.gd:
```diff
@@ -0,0 +1,214 @@
+
+
+func get_despawn_point() -> RoutePointDespawnPoint:
+	return route_points.back()  # cannot duplicate() objects :(
```
Okay, you've convinced me, I'll make them all Resources :smile:

DasCapschen commented on src/Data/Scripts/Resources/track_scenario.gd:
```diff
@@ -0,0 +1,44 @@
+
+# Dict[String, ScenarioRoute]
+# see World.gd spawn_train()
+export var routes := {}
```
I think this didn't work, because "ScenarioRoute is not a ResourceType". I believe this happens even if it `extends Resource`, only built-in types can be exported like that.

DasCapschen commented on src/Data/Scripts/World.gd:
```diff
@@ -140,69 +117,61 @@ func get_signal_scenario_data() -> Dictionary:
-
 func set_scenario_to_world() -> void:
-	scenario_manager.set_save_path(Root.current_scenario)
+	current_scenario = TrackScenario.load_scenario()
```
The player selects it in the play menu.
`TrackScenario.load_scenario()` loads `Root.current_scenario` by default, if not passed a path to a scenario file.

DasCapschen commented on src/Data/Scripts/World.gd:
```diff
@@ -8,7 +8,7 @@ var default_persons_at_station: int = 20
 
 var globalDict := {} ## Used, if some nodes need to communicate globally. Modders could use it. Please make sure, that you pick an unique key_name
```
Yes, but can we do it in a different PR? 😅

DasCapschen commented on src/Data/Scripts/chunk.gd:
```diff
@@ -0,0 +1,52 @@
+export (Array, String) var rails := []
+
+
+# see chunk_prefab.tscn
+# TrackObjects are saved in global coordinates, NOT relative to the chunk!
+# this is because they are attached to the Rails!
+# Buildings are also stored in global coordinates
+# DefaultGrass is stored relative to the chunk, we must move it to the correct global position
+# thus, Chunks should always be at (0,0,0)
```
See the next note, it requires a TrackObject refactor, afaik. I agree, but could we do it in a different PR?  
The way chunks work now (before tscn) is also in global coordinates, because everything gets loaded in as an individual object.  
So let's try and make it work the old way in the new format, and then refactor it to make it better :)

DasCapschen commented on src/Editor/Scripts/ScenarioConfiguration.gd:
```diff
@@ -78,58 +68,39 @@ func show_selection_message(text: String) -> void:
-	j_save_module.save_value("routes", routes)
-	j_save_module.reload()
-	routes = j_save_module.get_value("routes", {})
+	routes[duplicated_entry_name] = routes[source_entry_name]
```
You might be right there 🤔 
Will solve by using Resource type and .duplicate() :)

DasCapschen commented on src/Editor/Docks/Configuration/Configuration.gd:
```diff
@@ -1,13 +1,11 @@
-	save_path = find_parent("Editor").current_track_path + ".trackinfo"
-	j_save_module.set_save_path(save_path)
+	save_path = find_parent("Editor").current_track_path + "_config.tres"
+	world_config = load(save_path) as WorldConfig
```
Are you sure? It is loaded from a file, doesn't it need to know the type then? 🤔

HaSa1002 replied:
yeah but only when we set the variable type in the same call

HaSa1002 replied:
the type is defined earlier in this case

DasCapschen commented on src/Data/Scripts/DataClasses/scenario_route.gd:
```diff
@@ -0,0 +1,214 @@
+
+	if start_point is RoutePointSpawnPoint:
+		start_point.forward = calculated_rail_route[0].forward
+		return start_point as RoutePointSpawnPoint
```
Nope, actually necessary because the function returns RoutePointSpawnPoint. Without the cast, Godot complains that `the returned type (RoutePoint) doesn't match the functions return type (RoutePointSpawnPoint)` :)

DasCapschen commented:
I believe this feature is now finished.  
Feel free to review the code and more importantly, test it out in-game 😄 
@nalquas @HaSa1002

DasCapschen commented on src/Data/Scripts/chunk.gd:
```diff
@@ -0,0 +1,52 @@
+extends Spatial
+
+export var is_empty := true
+export var generate_grass := true  # set this to false when we have terrain
```
This does indeed sound like a good idea, but again, I think it is worth it's own PR. Do we already have an Issue for this feature?

nalquas replied:
> This does indeed sound like a good idea, but again, I think it is worth it's own PR. Do we already have an Issue for this feature?

I've created an issue: https://github.com/Libre-TrainSim/Libre-TrainSim/issues/399

nalquas commented:
Currently (at bb1395dee523dfeb3c3a40874f1c937f1b848300), opening any track in the track editor crashes the game with an `Assertion failed.` in `res://Data/Scripts/World.gd:46 at function: _ready`, as well as the following two errors in the Debugger:
```
E 0:00:27.686   load_interactive: Cannot open file 'user://editor/tutorial/Worlds/tutorial/tutorial_config.tres'.
  <C++ Error>   Condition "err != OK" is true. Returned: Ref<ResourceInteractiveLoader>()
  <C++ Source>  scene/resources/resource_format_text.cpp:1183 @ load_interactive()
  <Stack Trace> World.gd:45 @ _ready()
                Editor.gd:572 @ load_world()
                Editor.gd:23 @ _ready()
                loading_screen.gd:127 @ _add_to_tree()
                loading_screen.gd:104 @ _clean_up_and_switch()
                loading_screen.gd:74 @ _process()
```
```
E 0:00:27.686   _load: Failed loading resource: user://editor/tutorial/Worlds/tutorial/tutorial_config.tres. Make sure resources have been imported by opening the project in the editor at least once.
  <C++ Error>   Condition "found" is true. Returned: RES()
  <C++ Source>  core/io/resource_loader.cpp:270 @ _load()
  <Stack Trace> World.gd:45 @ _ready()
                Editor.gd:572 @ load_world()
                Editor.gd:23 @ _ready()
                loading_screen.gd:127 @ _add_to_tree()
                loading_screen.gd:104 @ _clean_up_and_switch()
                loading_screen.gd:74 @ _process()
```

DasCapschen commented:
That's interesting... I don't have any crashes, neither in Godot nor in export 🤔 
I will investigate why this happens.

HaSa1002 commented:
> Edit 2: Interestingly, the crash does not happen when I export the game and run that build. In the Godot Editor however, it crashes for me every time.

@nalquas it's an assert. It is not executed in exports. It's a sanity check which acts like a contract to allow us to program with minimal overhead under expectation certain requirements are met. Not meeting these requirements is a code writing time issue and thus a bug.

HaSa1002 commented:
The question really is: Is that assert a code writing time failure. Especially given it follows a load from the file system that may fail.
It may be a conversion issue but can you please share the content of the folder `user://editor/tutorial/Worlds/tutorial/`?

DasCapschen commented:
I suspect it somehow failed to save the `tutorial_config.tres`. 
It should be created when you convert an old map or create a new map in the Editor map selection screen.
Can you check your logs for further messages?

nalquas commented:
> @nalquas it's an assert. It is not executed in exports. It's a sanity check which acts like a contract to allow us to program with minimal overhead under expectation certain requirements are met. Not meeting these requirements is a code writing time issue and thus a bug.

You're right, I forgot (probably due to the time 🤷). Sorry


> The question really is: Is that assert a code writing time failure. Especially given it follows a load from the file system that may fail. It may be a conversion issue but can you please share the content of the folder `user://editor/tutorial/Worlds/tutorial/`?

Here you go: [tutorial.zip](https://github.com/Libre-TrainSim/Libre-TrainSim/files/9414600/tutorial.zip)


> I suspect it somehow failed to save the `tutorial_config.tres`. It should be created when you convert an old map or create a new map in the Editor map selection screen. Can you check your logs for further messages?

There aren't any further errors shown, however lots of warnings. I'm not sure which ones could be relevant, though, as there's nothing standing out of the random warnings. I can see a lot of `... returns a value, but this value is never used` warnings with quite a few of them coming from `name_node_appropriate` calls by `Editor.gd` or `RailBuilder.gd`. There's also some unused return values for `open()`, `make_dir_recursive()`, `change_dir()`, `pack()` and `save()`, with the former three coming from `export_track.gd` and the latter two coming from `Editor.gd`.
Aside from that, two variable shadowing warnings for `start_pos` and `end_pos` at `Rail.gd:478`.
Additionally, of course, there are the obligatory `Skipping pack loading in editor build, because of ...` warnings.
And lots of other random warnings.

So nothing really obviously relevant in the debugger or the logs.

DasCapschen commented:
This was exactly the problem, thank you. I just replicated it on my side too :)
I have fixed this by calling `_port_to_new_trackinfo()` before calling `load_world()`

Your tutorial world should now import correctly 👍🏻

nalquas commented:
> Your tutorial world should now import correctly 👍🏻

Thanks, that fixed it.

I have another crash with another (admittedly very empty and non-important) [test map](https://github.com/Libre-TrainSim/Libre-TrainSim/files/9419233/testonly.zip) though at `Editor.gd:230`: `Unable to iterate on object of type 'Nil'.` A simple check whether the old rail logic settings are empty should fix that, I think.

nalquas commented:
There also appears to be something wrong with the scenario conversion. I had two lines with different routes (stations) in the Tutorial map, however both lines now contain the stations of both original lines (Note the multiple First Stations, for example):

![Screenshot_20220824_223058](https://user-images.githubusercontent.com/22664233/186517465-2b293d2d-9c6a-4577-8fa9-cf0cb2c7c77d.png)

Simple enough to fix by hand of course, but still, thought I'd report it.

HaSa1002 commented:
Fair, I mean we are going to have fun converting Hainfurt and Nuremberg 😅

DasCapschen commented:
I've added a null check, give it a try when you can :)

DasCapschen commented:
> Fair, I mean we are going to have fun converting Hainfurt and Nuremberg

U2 is is almost done 😉 
There's just a missing tunnel somewhere I need to reconstruct 🙈

DasCapschen commented:
> There also appears to be something wrong with the scenario conversion. I had two lines with different routes (stations) in the Tutorial map, however both lines now contain the stations of both original lines (Note the multiple First Stations, for example):

I am aware of this bug, but I have no idea why it would happen or how to fix it.
It's a completely new function call for each scenario, somehow memory isn't cleared. 😦 
I think it might be jSaveModule.

nalquas commented:
> I've added a null check, give it a try when you can :)

Fixed that spot, but now there's another crash of the same type due to there also being no routes in the map:
`Unable to iterate on object of type 'Nil'.` at `Editor.gd:197` (`_convert_scenario`)

DasCapschen commented:
I am glad you are testing things I didn't even think of. Here I am assuming all old data is valid 😂 
Added a null check.

nalquas commented:
> I am glad you are testing things I didn't even think of. Here I am assuming all old data is valid :joy: Added a null check.

Minor issue with that fix: `Parser Error: A non-void function must return a value.` at `Editor.gd:263`

DasCapschen commented:
weird, Godot didn't complain to me about that 🤔 
fixed it.

Also, thank you very much for your work and review 🙏🏻

nalquas commented:
> weird, Godot didn't complain to me about that :thinking: fixed it.

Thanks, that worked. All my files are loadable now.

> Also, thank you very much for your work and review 🙏🏻

Glad to help :)

nalquas commented:
Actually, something else seems to be broken now. All of the objects in the map, aside from the rails, seem to have been moved into the middle of the map. Probably chunk-related? Maybe everything has been moved into the center chunk or something
![Screenshot_20220824_234700](https://user-images.githubusercontent.com/22664233/186529620-87dc98bb-1e9d-4ad7-86ac-c901f18872b2.png)

Here's the files for analysis:
[tutorial_after_conversion.zip](https://github.com/Libre-TrainSim/Libre-TrainSim/files/9419624/tutorial_after_conversion.zip)

nalquas commented on src/Data/Scripts/train_spawn_information.gd:
```diff
@@ -0,0 +1,11 @@
+var minimal_platform_length := 0
+var route := []
+var station_table := []
+var despawn_point
```
```suggestion
var despawn_point: RoutePointDespawnPoint
```
We could type this, I think

DasCapschen replied:
I think we can type it as `RoutePoint`, but we cannot type it as `RoutePointDespawnPoint`, because the despawn point can be either `RoutePointDespawnPoint` or `RoutePointStation`

DasCapschen commented:
And that's it. I've fixed the last few bugs in the Editor :)

... and doing so have completely re-written the Chunk loading. No more threads! 🥳 
Because we are using TSCN files now, I've made use of `ResourceLoader.load_interactive`, like the loading screen does.
Everything is on the main thread, but it's not blocking! And now it can actually be debugged!

I've made sure everything works in the Editor and U2 Nuremburg. Feel free to give it a try.
(Note: in the Editor chunks can be a bit slower to load, but I have not had any problems in-game)

nalquas commented on src/Editor/Modules/AddObjects.gd:
```diff
@@ -1,5 +1,11 @@
 extends Control
 
+var editor
+
+
+func _ready():
+	editor = find_parent("Editor")
+
```
I think this could be simplified using `onready`, no need to add the `_ready` function for just one variable.
```suggestion
onready var editor = find_parent("Editor")

```

nalquas commented:
Had a crash trying to save (Ctrl+S) in the scenario editor with no routes created:

`Invalid call. Nonexistent function 'duplicate' in base 'Nil'.` at `res://Editor/Scripts/ScenarioConfiguration.gd:44 - at function: save`

nalquas commented:
While in the scenario editor, a lot of the following error is being thrown:
```
E 0:03:37.103   decompose_polygon_in_convex: Convex decomposing failed!
  <C++ Source>  core/math/geometry.cpp:687 @ decompose_polygon_in_convex()
  <Stack Trace> Rail2DCollider.gd:56 @ _ready()
                scenario_map.gd:155 @ create_line2d_from_rail()
                scenario_map.gd:202 @ update_map()
                scenario_editor.gd:90 @ run_map_updater()
                scenario_editor.gd:27 @ _process()
```

nalquas commented:
Another quirk I noticed in an exported build of the game: Using the `Test track` button in the scenario editor fails with a message that you have to "create a scenario" to test, even though I've created one. Pressing `Test track` in the track editor, however, works flawlessly.
![Screenshot_20220828_140854](https://user-images.githubusercontent.com/22664233/187073449-1b97fd4e-4e5b-4c89-8150-8678703a766b.png)

And another thing: I noticed the "Waiting Persons" setting in the RoutePoints menu isn't limited to the 0-100% range, is this a bug or a feature?
![Screenshot_20220828_135540_cropped](https://user-images.githubusercontent.com/22664233/187073583-45a4b32e-6d1f-4a85-9b85-acbfdc1742e8.png)

DasCapschen commented:
> Had a crash trying to save (Ctrl+S) in the scenario editor with no routes created:

Nice one finding that 👍🏻 



> While in the scenario editor, a lot of the following error is being thrown:
> 
> ```
> E 0:03:37.103   decompose_polygon_in_convex: Convex decomposing failed!
>   <C++ Source>  core/math/geometry.cpp:687 @ decompose_polygon_in_convex()
>   <Stack Trace> Rail2DCollider.gd:56 @ _ready()
>                 scenario_map.gd:155 @ create_line2d_from_rail()
>                 scenario_map.gd:202 @ update_map()
>                 scenario_editor.gd:90 @ run_map_updater()
>                 scenario_editor.gd:27 @ _process()
> ```
> 
> I feel like that issue was already there before, though. It didn't affect anything visible to the user, anyway.
> 
> I don't think this has to be fixed in this PR, we could revisit this in a later PR.

This was already a problem. Something seems to be wrong with the code generating the colliders for the Rails. 
This also leads to the problem that some rails are not selectable.
I have replaced the single concave polygon by many rectangle shapes to approximate the line shape. Probably not the best solution, but is solves this problem for now.


> Another quirk I noticed in an exported build of the game: Using the `Test track` button in the scenario editor fails with a message that you have to "create a scenario" to test, even though I've created one. Pressing `Test track` in the track editor, however, works flawlessly. Probably not related to this PR though (?), we might want to create an issue to revisit later. 

This was already a bug. The scenario editor and the track editor used different paths trying to find the scenarios.
I have adjusted this behaviour, and it should work now.

> And another thing: I noticed the "Waiting Persons" setting in the RoutePoints menu isn't limited to the 0-100% range, is this a bug or a feature? Again, maybe not related to this PR.

I'm not sure about that one. I haven't worked on the PersonSystem. I don't even know why this number is in Percent to begin with. Probably best to leave it as is for now.

nalquas commented:
Thanks, fixes work for me.

However, something in b68391b3e513e74bf254c01308b241f21088e21c has resulted in the Editor not showing any objects except the track:

| Before (at 36945ef87ba891b124736e11dfdc60d10edfced3)      | After (at b68391b3e513e74bf254c01308b241f21088e21c) |
| ----------- | ----------- |
| ![Screenshot_20220828_162831](https://user-images.githubusercontent.com/22664233/187079112-6f6d69b5-5417-444b-a5d7-c4692c320fbb.png)      | ![Screenshot_20220828_162724](https://user-images.githubusercontent.com/22664233/187079118-74a14c3c-77b5-4bda-a900-779372ba399a.png)       |

The objects show as normal when testing the track, though.

DasCapschen commented:
Indeed, I missed some path changes. Did a global search and should've fixed them all now. 
Works on my machine, feel free to test again :)

nalquas commented:
Yup, that fixed it.

*However*, the "Test track" button in the scenario editor is now differently broken. When loading the track from here, I get a grey screen and nothing happens.

https://user-images.githubusercontent.com/22664233/187080788-2a1e47d2-a7e8-4195-b0e9-924c1c765ea1.mp4

Using "Test track" in the track editor still works fine, btw.

nalquas commented:
Sorry, I keep finding errors :(
Creating a new world/track and trying to save results in the following:

```
E 0:00:07.551   load_interactive: Cannot open file 'user://editor/testtracktest/Worlds/testtracktest/testtracktest_config.tres'.
  <C++ Error>   Condition "err != OK" is true. Returned: Ref<ResourceInteractiveLoader>()
  <C++ Source>  scene/resources/resource_format_text.cpp:1183 @ load_interactive()
  <Stack Trace> World.gd:46 @ _ready()
                Editor.gd:727 @ load_world()
                Editor.gd:28 @ _ready()
                loading_screen.gd:127 @ _add_to_tree()
                loading_screen.gd:104 @ _clean_up_and_switch()
                loading_screen.gd:74 @ _process()
```

DasCapschen commented:
Don't be sorry. Every found error helps 😄 
I just found the issue with the missing `_config.tres` myself and fixed it. 

Found the bug with the scenario editor. If you were exporting, it wasn't correctly setting `Root.scenario_editor = false` ... 
That's why global variables are bad, mkay.

nalquas reviewed (approved):
Everything seems to be working now. Looks good from my side. :+1: 
Good job and thanks for all the fixes

HaSa1002 commented:
The free driving scenario doesn't seem to end. It's winning condition is not triggered.

HaSa1002 commented:
Overriding a red signal prompts the failure message but the scene is not switched on button click

HaSa1002 commented:
Going over the end of the of the railroute after overriding the signal does terminate the play correctly.

HaSa1002 commented:
The beginner tutorials don't show any text messages.

HaSa1002 commented:
The saftety and signals tuorials do work but they don't have a speed set at the beginning

HaSa1002 commented:
On load of an old map, I get the following errors:
```
E 0:25:44.945   surface_set_material: Index p_idx = 1 is out of bounds (surfaces.size() = 1).
  <C++ Source>  scene/resources/mesh.cpp:924 @ surface_set_material()
  <Stack Trace> Editor.gd:158 @ _port_to_new_chunk_system()
                Editor.gd:33 @ _ready()
                loading_screen.gd:127 @ _add_to_tree()
                loading_screen.gd:104 @ _clean_up_and_switch()
                loading_screen.gd:74 @ _process()
```
and
![grafik](https://user-images.githubusercontent.com/14185889/188331929-1bc55d91-6ad9-42f2-bd43-069e9a856563.png)

HaSa1002 commented:
Clicking "New Rail" in the rail builder on a freshly created map, causes this wonderful bug:
![grafik](https://user-images.githubusercontent.com/14185889/188332054-87440375-3564-4858-aa99-c67884014388.png)
It should be `translation` instead of position

HaSa1002 commented:
I just retested the tutorials in the main branch and they just as broken as they are here. Do you wanna try to fix them or do we move that to the future?

HaSa1002 reviewed (approved):
lgtm

HaSa1002 commented:
Thanks ❤️
